File: Palantir challenge/Palantir.java
import java.util.*;

public class Palantir{

	public Palantir() {}

	public void parseExpression(List<String> expr){
		String infixDecimalExpression = "";
		for (String str : expr){
			if (!str.equals("+") && !str.equals("*") && !str.equals("-")){
				String temp = convertToBase10(str);
				infixDecimalExpression += temp;
			}
			else{
				infixDecimalExpression += str;
			}
		}

		String postfix = convertToPostfix(infixDecimalExpression);
		System.out.println(postfix);
		System.out.println(evaluatePostfix(postfix));
	}

	private int evaluatePostfix(String postfix){
		int result = 0;
		Deque<Integer> s = new ArrayDeque<Integer>();
		String[] tokens = postfix.split(" ");
		for (int i = 0; i < tokens.length ; i++ ){
			if (!"+".equals(tokens[i]) && !"*".equals(tokens[i]) && !"-".equals(tokens[i])){
				//System.out.println("pushed to stack: " + tokens[i]);
				s.push(Integer.parseInt(tokens[i]));
			}
			else{
				int number1 = s.pop();
                int number2 = s.pop();
                //System.out.println("taken 2 numbers: " + number1 + " and " + number2);
                switch(tokens[i]) {
                    case "+":
	                    result = number1 + number2;
	                    break;
                    case "-":
	                    result = number2 - number1;
	                    break;
                    case "*":
	                    result = number1 * number2;
	                    break;  
                } 
                s.push(result);
                //System.out.println("pushed to stack result: " + result);
			}
		}
		return s.pop();
	}

	private int precedence(char e){
		if (e == '+' || e == '-')
			return 1;
		else
			return 2;
	}

	private String convertToPostfix(String infix){

		String postfix = "";
		Deque<Character> deque = new ArrayDeque<Character>();

		for (int i = 0; i < infix.length(); i++){
			if (Character.isDigit(infix.charAt(i))) {
				postfix += infix.charAt(i);
			}
			else{
				postfix += ' ';
				if (deque.isEmpty())
					deque.push(infix.charAt(i));
				else{
					while (!deque.isEmpty() && precedence(infix.charAt(i)) <= precedence(deque.peekFirst())) {
						postfix += deque.pop();
						postfix += ' ';
					}
					deque.push(infix.charAt(i));
				}
			}
		}

		while (deque.size() > 0){
			postfix += ' ';
			postfix += deque.pop();
			postfix += ' ';
		}
		return postfix;
	}

	private String convertToBase10(String num){
		int result = 0;
		String[] pNumber = num.split("_");
		String number = pNumber[0];
		int base = Integer.parseInt(pNumber[1]);
		for (int i = 0; i < number.length(); i++){
			result +=  Character.getNumericValue(number.charAt(number.length() - i - 1)) * Math.pow(base, i);
		}
		return "" + result;
	}

	public static void main(String[] args){
		List<String> tokens = new ArrayList<String>();
		Palantir p;
		Scanner sc = new Scanner(System.in);
		while(sc.hasNext()){
			String expr = sc.next();
			tokens.add(expr);
		}
		p = new Palantir();
		p.parseExpression(tokens);
	}
}

File: Palantir challenge/Palantir.rb
class Palantir

	def parse_expression(expr) 
		infix = ""
		expr = expr.split(" ")
		s = ["+", "-", "*"]
		expr.map do |e|
			if !s.include?(e)
				e = e.split("_")
				infix += convert_to_base_10(e[0], e[1])
			else
				infix += e
			end
		end
		#p convert_to_postfix(infix)
		p evaluate_postfix (convert_to_postfix(infix))
	end

	def convert_to_base_10(num, base)
		base_ = {"A" => 10, "B" => 11, "C" => 12, "D" => 13, "E" => 14, "F" => 15}
		result = 0
		i = 1
		num.split("").each do |char|
			if base.to_i < 10
				result += char.to_i * (base.to_i ** (num.length-i))
				i += 1
			else
				if base_.include?(char)
					result += base_[char] * (base.to_i ** (num.length-i))
				else
					result += char.to_i * (base.to_i ** (num.length-i))
				end
				i += 1
			end
		end
		result.to_s
	end

	def evaluate_postfix postfix
		result = 0
		stack = []
		postfix.split(" ").each do |char|
			if /^[0-9]/ === char
				stack.push(char)
			else
				n1 = stack.pop
				n2 = stack.pop
				case char
				when "+"
					result = n1.to_i + n2.to_i
				when "-"
					result = n2.to_i - n1.to_i
				when "*"
					result = n1.to_i * n2.to_i
				end
				stack.push(result)
			end
		end
		stack.pop
	end

	def precedence a
		prec = {"+" => 1, "-" => 1, "*" => 2}
		prec[a]
	end

	def convert_to_postfix infix
		stack = []
		postfix = ""
		infix.split("").each do |char|
			if /^[0-9]/ === char		#if char is a number
				postfix += char 		#add to postfix		
			else
				postfix += " "
				if stack.empty?
					stack.push(char)
				else
					while !stack.empty? && (precedence(stack.last) >= precedence(char))
						postfix += stack.pop
						postfix += " "
					end
					stack.push(char)
				end
			end
		end
		while !stack.empty?
			postfix += " "
			postfix += stack.pop
			postfix += " "
		end
		postfix
	end

end

#input = gets
#input = input.chomp

palantir = Palantir.new
palantir.parse_expression("12_10 + 111_2 + 1ABC_16")
palantir.parse_expression("12_10 + 111_2 * 11_10")
palantir.parse_expression("1430_5 - 110_2 * 2A_12 + 10_10")
